addons/binding/org.openhab.binding.xiaomivacuum/src/main/java/org/openhab/binding/xiaomivacuum/discovery/VacuumDiscovery.java

```diff
@@ -0,0 +1,196 @@
+                        id);
+                thingDiscovered(DiscoveryResultBuilder.create(uid)
+                        .withProperty(XiaomiVacuumBindingConstants.PROPERTY_HOST_IP, i.getKey())
+                        .withRepresentationProperty(id).withLabel("Discovered Xiaomi Mi IO Device").build());
+            } else {
+                logger.debug("Discovered token for device {}: {} ('{}')", id, token, new String(msg.getChecksum()));
+                thingDiscovered(DiscoveryResultBuilder.create(uid)
+                        .withProperty(XiaomiVacuumBindingConstants.PROPERTY_HOST_IP, i.getKey())
+                        .withProperty(XiaomiVacuumBindingConstants.PROPERTY_TOKEN, token).withRepresentationProperty(id)
+                        .withLabel("Discovered Xiaomi Mi IO Device").build());
+            }
+        }
+    }
+
+    /**
+     * @return broadcast addresses for all interfaces
+     */
+    private TreeSet<String> getBroadcastAddresses() {
+        TreeSet<String> broadcastAddresses = new TreeSet<String>();
+        try {
+            broadcastAddresses.add("224.0.0.1");
+            broadcastAddresses.add("192.168.3.109");
+            Enumeration<NetworkInterface> interfaces = NetworkInterface.getNetworkInterfaces();
+            while (interfaces.hasMoreElements()) {
+                try {
+                    NetworkInterface networkInterface = interfaces.nextElement();
+                    if (networkInterface.isLoopback() || !networkInterface.isUp()) {
+                        continue;
+                    }
+                    for (InterfaceAddress interfaceAddress : networkInterface.getInterfaceAddresses()) {
+                        String address = interfaceAddress.getBroadcast().getHostAddress();
+                        if (address != null) {
+                            broadcastAddresses.add(address);
+                        }
+                    }
+                } catch (Exception e) {
+                    // ignore
+                }
+            }
+        } catch (Exception e) {
+            logger.trace("Error collecting broadcast addresses: {}", e.getMessage(), e);
+        }
+        return broadcastAddresses;
+    }
+
+    public HashMap<String, byte[]> sendDiscoveryRequest(String ipAddress) {
+        HashMap<String, byte[]> responses = new HashMap<String, byte[]>();
+
+        try (DatagramSocket clientSocket = new DatagramSocket()) {
+            clientSocket.setReuseAddress(true);
+            clientSocket.setBroadcast(true);
+            clientSocket.setSoTimeout(TIMEOUT);
+            byte[] sendData = XiaomiVacuumBindingConstants.DISCOVER_STRING;
+            DatagramPacket sendPacket = new DatagramPacket(sendData, sendData.length, InetAddress.getByName(ipAddress),
+                    XiaomiVacuumBindingConstants.PORT);
+            for (int i = 1; i <= 2; i++) {
```

Comment:
the other examples I found send 3 times.  We need to see if that is really required.